web-server/src/grading-queue/create-or-update.ts

```diff
@@ -0,0 +1,103 @@
+import { RedisClientType } from "redis";
+import { GradingJob, EnrichedGradingJob } from "./types";
+import { createQueueKey, generateNonce } from "./utils";
+
+export const createOrUpdateGradingJob = async (
+  client: RedisClientType,
+  job: GradingJob,
+  isImmediateJob: boolean = false,
+) => {
+  const queueKey = createQueueKey(job.key, job.response_url);
+  const jobExists = await client.EXISTS(queueKey);
+  if (jobExists) {
+    const existingJob = JSON.parse(
+      (await client.GET(job.key)) as string,
+    ) as EnrichedGradingJob;
+    client.SET(
+      queueKey,
+      JSON.stringify({
+        ...job,
+        created_at: existingJob.created_at,
+        release_at: existingJob.release_at,
+      }),
+    );
+  } else {
+    const arrivalTime = Date.now();
+    const releaseTime = arrivalTime + job.priority;
+    await enqueueJob(client, job, arrivalTime, queueKey, isImmediateJob);
+    await client.SET(
+      job.key,
+      JSON.stringify({
+        ...job,
+        created_at: arrivalTime,
+        release_at: releaseTime,
+      }),
+    );
+  }
+};
+
+const enqueueJob = async (
+  client: RedisClientType,
+  job: GradingJob,
+  arrivalTime: number,
+  orcaKey: string,
+  isImmediateJob: boolean,
+) => {
+  if (isImmediateJob) {
+    await createImmediateJob(client, job, arrivalTime, orcaKey);
+  } else {
+    await createJob(client, job, arrivalTime, orcaKey);
+  }
+};
+
+const createJob = async (
+  client: RedisClientType,
+  gradingJob: GradingJob,
+  arrivalTime: number,
+  orcaKey: string,
+): Promise<void> => {
+  const { priority, collation } = gradingJob;
+  const releaseTime = arrivalTime + priority;
```

Comment:
I don't know if we actually document whether priority is in milliseconds of seconds, but it _needs_ to be, and needs to be reflected in this logic.

Reply:
This occurs in *numerous* locations; ensure these are all addressed.